src/common/plugin/fulltext/elasticsearch/ESAdapter.cpp

```diff
@@ -0,0 +1,280 @@
+  folly::dynamic body = folly::dynamic::object();
+  body["query"] = folly::dynamic::object();
+  body["query"]["regexp"] = folly::dynamic::object("text", pattern);
+  return ESAdapter::query(index, body);
+}
+
+StatusOr<ESQueryResult> ESAdapter::fuzzy(const std::string& index,
+                                         const std::string& pattern,
+                                         const std::string& fuzziness) {
+  folly::dynamic body = folly::dynamic::object();
+  body["query"] = folly::dynamic::object();
+  body["query"]["fuzzy"] = folly::dynamic::object();
+  body["query"]["fuzzy"]["text"] = folly::dynamic::object("value", pattern)("fuzziness", fuzziness);
+  return ESAdapter::query(index, body);
+}
+
+// StatusOr<ESQueryResult> ESAdapter::term(const std::string& index,const std::vector<std::string>&
+// words){
+//   folly::dynamic query = folly::dynamic::object("query",folly::dynam)
+// }
+
+StatusOr<ESQueryResult> ESAdapter::match_all(const std::string& index) {
+  folly::dynamic body = folly::dynamic::object();
+  body["query"] = folly::dynamic::object();
+  body["query"]["match_all"] = folly::dynamic::object();
+  return ESAdapter::query(index, body);
+}
+
+StatusOr<ESQueryResult> ESAdapter::query(const std::string& index, const folly::dynamic& query) {
+  auto result = randomClient().search(index, query);
+  if (!result.ok()) {
+    return std::move(result).status();
+  }
+  auto resp = std::move(result).value();
+  if (resp.count("hits")) {
+    auto hits = resp["hits"]["hits"];
+    ESQueryResult res;
+    for (auto& hit : hits) {
+      auto source = hit["_source"];
+      ESQueryResult::Item item;
+      item.text = source["text"].getString();
+      item.src = source["src"].getString();
+      item.dst = source["dst"].getString();
+      item.rank = source["rank"].getInt();
+      item.vid = source["vid"].getString();
+      res.items.emplace_back(std::move(item));
+    }
+    return res;
+  }
+  auto error = resp["error"];
+  if (error.isObject()) {
+    return Status::Error(folly::toJson(error));
+  }
+  return Status::Error(folly::toJson(resp));
+}
+
+std::string ESAdapter::genDocID(const std::string& vid,
+                                const std::string& src,
+                                const std::string& dst,
+                                int64_t rank) {
+  std::string str;
+  unsigned char mdStr[32] = {0};
+  if (!vid.empty()) {
+    str = vid;
+  } else {
+    str = src + dst + std::to_string(rank);
+  }
+  SHA256(reinterpret_cast<unsigned char*>(str.data()), str.size(), mdStr);
+  char hex[64] = {0};
+  for (int i = 0; i < 32; i++) {
+    hex[i * 2] = 'A' + mdStr[i] / 16;
+    hex[i * 2 + 1] = 'A' + mdStr[i] % 16;
+  }
+  return std::string(hex, 64);
+}
+
+ESClient& ESAdapter::randomClient() {
+  static thread_local std::default_random_engine engine;
+  static thread_local std::uniform_int_distribution<size_t> d(0, clients_.size() - 1);
+  return clients_[d(engine)];
```

Comment:
I don't quite remember, do we already support sign in multiple es client?

Reply:
I did not verify it, but it is supported in grammar and implementation.

Reply:
👌